### PR TITLE
Fix running on Spark 

### PR DESCRIPTION
**Pull request by dvadym**

### CODE REVIEW

File: pipeline_dp/dp_engine.py [hunk 1]
```diff
@@ -176,11 +178,35 @@ def filter_fn(captures: Tuple[MechanismSpec, int],
             partitions to keep."""
             mechanism, max_partitions = captures
             accumulator = row[1]
-            partition_selection_strategy = create_truncated_geometric_partition_strategy(
+            partition_selection_strategy = partition_selection.create_truncated_geometric_partition_strategy(
```
Comment (dvadym) [inline pipeline_dp/dp_engine.py:181]: This fixes PyCapsule serialization error

File: pipeline_dp/dp_engine.py [hunk 2]
```diff
@@ -90,6 +90,8 @@ def aggregate(self, col, params: AggregateParams,
             pass
         # col : (partition_key, accumulator)
 
+        col = self._fix_budget_accounting_for_spark(col, accumulator_factory)
```
Comment (boyaryn) [inline pipeline_dp/dp_engine.py:93]: It's great that you've solved the serialization problem for Spark.
My only concern is this line seems to violate some object-oriented design principle, probably D from SOLID. Should we redesign this? Or should it be left as is now and hopefully refactored later (I see this implementation is still in development).

**Review by boyaryn — COMMENTED** [review]
In general, looks great!

File: pipeline_dp/dp_engine.py (hunk 2 above)
Comment (dvadym) [reply pipeline_dp/dp_engine.py]: yeah, it makes sense. I agree that code here shouldn't know that it runs on Spark.

On other hand, PipelineOperations class knows nothing about accumulators and budgets, it know only about framework details. So updates of the accumulators should be initiated on higher level, and `DPEngine` looks the right place.

I've renamed this function to `_fix_budget_accounting_if_needed` and renamed corresponding PipelineOperation function to `is_serialization_immediate_on_reduce_by_key`. Such way we avoid dependency on knowing the framework.

WDYT?

Comment (boyaryn) [reply pipeline_dp/dp_engine.py]: IMHO the new version of the code looks much better with this application of the strategy pattern.

Thanks, approved.

**Review by boyaryn — APPROVED** [review]

File: pipeline_dp/dp_engine.py (hunk 2 above)
Comment (dvadym) [reply pipeline_dp/dp_engine.py]: Thanks for review!

**Pull request merged**